homeassistant/components/zha/__init__.py

```diff
@@ -205,35 +194,37 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
     websocket_api.async_load_api(hass)
 
     async def async_shutdown(_: Event) -> None:
-        await zha_gateway.shutdown()
+        """Handle shutdown tasks."""
+        assert ha_zha_data.gateway_proxy is not None
+        await ha_zha_data.gateway_proxy.shutdown()
 
     config_entry.async_on_unload(
         hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_shutdown)
     )
 
-    await zha_gateway.async_initialize_devices_and_entities()
+    await ha_zha_data.gateway_proxy.async_initialize_devices_and_entities()
     await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
     async_dispatcher_send(hass, SIGNAL_ADD_ENTITIES)
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
     """Unload ZHA config entry."""
-    zha_data = get_zha_data(hass)
+    ha_zha_data = get_zha_data(hass)
+    ha_zha_data.config_entry = None
 
-    if zha_data.gateway is not None:
-        await zha_data.gateway.shutdown()
-        zha_data.gateway = None
+    if ha_zha_data.gateway_proxy is not None:
+        await ha_zha_data.gateway_proxy.shutdown()
+        ha_zha_data.gateway_proxy = None
 
     # clean up any remaining entity metadata
     # (entities that have been discovered but not yet added to HA)
     # suppress KeyError because we don't know what state we may
     # be in when we get here in failure cases
     with contextlib.suppress(KeyError):
         for platform in PLATFORMS:
-            del zha_data.platforms[platform]
+            del ha_zha_data.platforms[platform]
 
-    GROUP_PROBE.cleanup()
     websocket_api.async_unload_api(hass)
 
     # our components don't have unload methods so no need to look at return values
```

Comment:
Does this change behavior in any way? I can add this change but `return True` was in ZHA since before the rewrite (https://github.com/home-assistant/core/blob/5f088e0501403a546df70167bcfd082852e94b50/homeassistant/components/zha/__init__.py#L239-L242) and I'd rather keep this single PR as small as it can be. We can always address it in a followup PR.